File: Ajedrez/app/src/main/java/com/example/ajedrez/View/Lessons/LessonsListFragment.java
package com.example.ajedrez.View.Lessons;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.ajedrez.Model.Assistance;
import com.example.ajedrez.Model.Lesson;
import com.example.ajedrez.R;
import com.example.ajedrez.View.MainActivity;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

public class LessonsListFragment extends Fragment {

    private LessonsListener mListener;
    private RecyclerView recyclerView;
    private LessonsListAdapter adapter;
    private List<Lesson> lessonList;

    public void setListener (MainActivity activity) {
        this.mListener = activity;
    }

    public static LessonsListFragment newInstance() {
        return new LessonsListFragment();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_lessons_list, container, false);
        recyclerView = view.findViewById(R.id.lessonsList);
        lessonList = new ArrayList<>();
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        adapter = new LessonsListAdapter(lessonList, mListener);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
        recyclerView.setAdapter(adapter);
        loadAssistance();
    }

    private void loadAssistance() {
        Query studentsQuery = FirebaseDatabase.getInstance().getReference().child("assistance");
        studentsQuery.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                List<Lesson> lessonsList = new ArrayList<>();
                for(DataSnapshot assistanceList :dataSnapshot.getChildren()){
                    String date = assistanceList.getKey();
                    List<Assistance> assistants = new ArrayList<>();
                    for(DataSnapshot student:assistanceList.getChildren()){
                        Assistance value = student.getValue(Assistance.class);
                        assistants.add(value);
                    }
                    lessonsList.add(new Lesson(date,assistants));
                }
                adapter.setLessonsList(lessonsList);
                adapter.notifyDataSetChanged();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    public interface LessonsListener {
    }
}

File: Ajedrez/app/src/main/java/com/example/ajedrez/view/lessons/LessonsListFragment.kt
package com.example.ajedrez.view.lessons

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.ajedrez.model.Lesson
import com.example.ajedrez.network.Network
import com.example.ajedrez.R
import com.example.ajedrez.view.BaseFragment
import java.util.*

class LessonsListFragment : BaseFragment(), Observer {

    //==============================================================================================
    // View objects
    //==============================================================================================

    private var mListener: LessonsListener? = null
    private var recyclerView: RecyclerView? = null
    private var adapter: LessonsListAdapter? = null
    private var lessonList: List<Lesson>? = null

    //==============================================================================================
    // Lifecycle
    //==============================================================================================

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_lessons_list, container, false)
        recyclerView = view.findViewById(R.id.lessonsList)
        lessonList = Network.getInstance().lessons
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        adapter = lessonList?.let { LessonsListAdapter(it, mListener) }
        recyclerView!!.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        recyclerView!!.adapter = adapter
        loadAssistance()
    }

    override fun onResume() {
        super.onResume()
        hideKeyboardFrom(context, view)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            mListener = context as LessonsListener
        } catch (e: ClassCastException) {
            throw ClassCastException("$context must implement MyInterface ")
        }
    }

    //==============================================================================================
    // Methods
    //==============================================================================================

    private fun loadAssistance() {
        lessonList?.let { adapter!!.setLessonsList(it) }
        adapter!!.notifyDataSetChanged()
    }

    companion object {
        fun newInstance(): LessonsListFragment {
            return LessonsListFragment()
        }
    }

    //==============================================================================================
    // LessonNotifications
    //==============================================================================================

    override fun update(p0: Observable?, p1: Any?) {
        lessonList = Network.getInstance().lessons
        loadAssistance()
    }
}